### ISSUE TITLE
Allow custom shell in Console tab

### ISSUE THREAD

**Issue by asherber** [gitextensions/gitextensions#7523]
## Feature description

In the Settings dialog, I currently have a choice of bash, cmd, or powershell as the shell to run in the Console tab of the Browse dialog.

It would be nice if there were a way for us to specify the path to whatever shell we wanted, including command line arguments.

(I also think that PowerShell Core should be added as one of the presets.)

**Comment by RussKie** [comment]
> In the Settings dialog, I currently have a choice of bash, cmd, or powershell as the shell to run in the Console tab of the Browse dialog.

Could you please explain what you mean? Screenshots would work too.
Because I'm not aware of such functionality in GE. We have a feature request #6460, but it is still open...

**Comment by pmiossec** [comment]
>Because I'm not aware of such functionality in GE. 
See #3403 and more especially for screenshot: https://github.com/gitextensions/gitextensions/pull/3403#issuecomment-265326058

>We have a feature request #6460, but it is still open...

This request is to be able to use another shell when opening a console window from the toolbar

**Comment by asherber** [comment]
Here's the current settings page. This controls which shell gets used in the Console tab of the Browse dialog.

![image](https://user-images.githubusercontent.com/5248041/71089258-70c06e80-216e-11ea-8a98-877a14738db9.png)

As @pmiossec says, #6460 is about the toolbar button, not the Console tab. (Though I suspect it would be nice for the toolbar button to have more options too, and possibly to use the same shell specified for the Console tab. But that's for another time!)

**Comment by pmiossec** [comment]
>(I also think that PowerShell Core should be added as one of the presets.)

After a really quick look, the `pwsh.exe` file is not in the `PATH` on my computer.
So it will be difficult to support it by default.

A custom shell is a better solution.

**Comment by asherber** [comment]
Silly question, but did you install PowerShell Core? It's on my path after installation.

For vanilla installations, the default install dir is also `c:\Program Files\PowerShell\6\pwsh.exe`, where `6` is the current major version number. GE could probe to determine latest version installed.

A custom shell lets the user do whatever he wants, which is good; additional presets simplify the user experience for the most common cases.

**Comment by pmiossec** [comment]
>Silly question, but did you install PowerShell Core? It's on my path after installation.

Yes, I don't remember when (and so I don't how I did it). I've got it in `C:\Program Files\PowerShell\6.0.2` but this folder is not in may PATH and I don't know if I am the exception or if it's the normal case. And that's the problem if we want to add it as an option.

That is strange because when I do `where pwsh`, it return nothing because it is not in the PATH but I still am able to run it with `Win+r` and `pwsh`. So the system should know it form one way or the other.

>For vanilla installations, the default install dir is also c:\Program Files\PowerShell\6\pwsh.exe

As you could notice, that's not exactly the case for me.

>A custom shell lets the user do whatever he wants, which is good; additional presets simplify the user experience for the most common cases.

Yes, I well understood that 😉

**Comment by RussKie** [comment]
LOL I feel dumb, I didn't know we had it there... We have so many options
and configurations!

We could it in a number of ways.

1. Configuration
Add a 'custom' entry that would allow users to point to a desired tool.
This begs the question whether we need the presets at all. Probably useful,
but good to consider.

2. The console tab can be refactored/enhanced to resemble how the console
is implemented in VSCode, e.g. have a dropdown and allow users to
spawn/kill different sessions with different tools.

Wrt the config - I want it to have redone with a PropertyGrod control.

**Comment by ivangrek** [comment]
Is it relevant to do?

**Comment by pmiossec** [comment]
I don't know if it's relevant because no one upvote the request, so I think that's not hugely requested and, personally, I don't really feel the need for it.
But yes, that's still something that could be done 😉

**Comment by ivangrek** [comment]
I would like to work on this issue.
Is it correct to think of any executable file as a shell?

**Comment by pmiossec** [comment]
> I would like to work on this issue.

👍 

> Is it correct to think of any executable file as a shell?

Yes. I think we could assume that it is the case. I can't imagine something that could allow us to find if an executable is a shell or not. We let the responsibility to the user...

**Comment by asherber** [comment]
I would suggest that the UI needs three controls. A dropdown would let the user pick one of our presets, or Custom; if Custom is selected, then two edit boxes let the user specify path and arguments.

**Comment by pmiossec** [comment]
>I would suggest that the UI needs three controls. A dropdown would let the user pick one of our presets, or Custom; if Custom is selected, then two edit boxes let the user specify path and arguments.

That's the most straightforward possibility.

If you are more ambitious, I have another one 😉 

1. Being able to configure a list of custom shells (with the path and arguments)
2. Being able to choose one of the bundled or custom shells as the default one that will be run in the console
3. Display all these custom shells after the bundled ones in the toolbar (with a separator): 
![image](https://user-images.githubusercontent.com/460196/92388615-e1058a80-f117-11ea-963a-e8c220f34afa.png)

**Comment by RussKie** [comment]
> If you are more ambitious, I have another one 😉
> 
> 1. Being able to configure a list of custom shells (with the path and arguments)
> 2. Being able to choose one of the bundled or custom shells as the default one that will be run in the console
> 3. Display all these custom shells after the bundled ones in the toolbar (with a separator):

I'd start with something simple, so we can gauge the usefulness of the feature. I'd hate to spend a lot of time and effort building something users won't be useful.
So perhaps start with launching the configured shell.
Or go the VS Code route and provide a dropdown so a user could start a new console session (and kill an existing)
![image](https://user-images.githubusercontent.com/4403806/92424152-df46d000-f1c6-11ea-844b-89bcf042c2fe.png)

It would also be great to plug in Windows Terminal.

> If you are more ambitious, I have another one 😉
> 
> 1. Being able to configure a list of custom shells (with the path and arguments)

If we go down this route, it may worth aligning a proposed configuration with configurations/syntax of VS Code and/or WSL, which users likely be familiar with.


A little OT, though related - we may wish to enhance our toolbar dropdown at least with tooltips:
![image](https://user-images.githubusercontent.com/4403806/92424198-03a2ac80-f1c7-11ea-9ef8-9d6de8e75173.png)

**Comment by ivangrek** [comment]
> So perhaps start with launching the configured shell.
Or go the VS Code route and provide a dropdown so a user could start a new console session (and kill an existing)

https://github.com/gitextensions/gitextensions/issues/7523#issuecomment-688299101

> 2. Being able to choose one of the bundled or custom shells as the default one that will be run in the console
> 3. Display all these custom shells after the bundled ones in the toolbar (with a separator):

Are you suggesting to remove the ability to run the shell in a separate window and implement this feature on the tab?

> If we go down this route, it may worth aligning a proposed configuration with configurations/syntax of VS Code and/or WSL, which users likely be familiar with.

And not use the settings window?

Could you please describe the desired final functionality beyond the ability to add a shell.

**Comment by RussKie** [comment]
The dropdown is one of the way to launch a console, which is being configured via Settings > Browser repository windows (being updated in https://github.com/gitextensions/gitextensions/pull/8307).
And we should apply the same to the Console tab - i.e. the Console tab should start with the configured default shell, not bash as it does now.

My braindump in https://github.com/gitextensions/gitextensions/issues/7523#issuecomment-688569425 was about allowing to launch a new console within the Console tab in addition (or replacing) the default shell. Such experience can be implemented in a similar way to VS Code. I added this screenshot to help visualise it:
![image](https://user-images.githubusercontent.com/4403806/92424152-df46d000-f1c6-11ea-844b-89bcf042c2fe.png)
**This is a stretch goal.**

The 2nd screenshot was an observation, to make our toolstrip dropdown a little more informative - i.e. provide more information on shell we're about to launch. This maybe a quick win.

**Comment by ivangrek** [comment]
> 3. Display all these custom shells after the bundled ones in the toolbar (with a separator)

Is everyone on the toolbar or in the dropdown?

Second commit:

![image](https://user-images.githubusercontent.com/3169265/92724141-d135b280-f372-11ea-9aff-f3aae62765d9.png)

---
`AppSettings.Saved += FillUserShells;`

I think we need something like this

AppSettings.[Value].Updated += FillUserShells;

Only if [New Value]  != [Old Value]

**Comment by ivangrek** [comment]
Third commit:

![terminal](https://user-images.githubusercontent.com/3169265/92733236-20361480-f380-11ea-9eb8-8889b496c662.gif)

**Comment by ivangrek** [comment]
Need help.
I am trying to close my terminal session and it opens a console application.

![image](https://user-images.githubusercontent.com/3169265/92757608-904f9500-f396-11ea-80ab-05bc62bbb997.png)

**Comment by RussKie** [comment]
> Is everyone on the toolbar or in the dropdown?

Let's start with the toolbar for now. When we get to implementing a dropdown on the Console tab we can add them there too.
